test/integration/http2_flood_integration_test.cc

```diff
@@ -1564,4 +1587,582 @@ TEST_P(Http2FloodMitigationTest, UpstreamFloodDetectionIsOnByDefault) {
+    EXPECT_GE(3 * connection_watermark, buffer_factory_->maxBufferSize());
+    auto overflow_info = buffer_factory_->maxOverflowRatio();
+    EXPECT_LT(1.0, std::get<0>(overflow_info));
+    EXPECT_GE(3.0, std::get<0>(overflow_info));
+    // Max overflow happens on the connection buffer.
+    EXPECT_EQ(connection_watermark, std::get<1>(overflow_info));
+  }
+  EXPECT_THAT(buffer_factory_->highWatermarkRange(), testing::Pair(32768, 128 * 1024));
+}
+
+// Verify that buffering is limited when handling small responses.
+TEST_P(Http2BufferWatermarksTest, SmallResponseBuffering) {
+  int num_requests = 200;
+  uint64_t kNumBlocks = 1;
+  uint64_t kBlockSize = 10240;
+  uint32_t connection_watermark = 32768;
+  uint32_t downstream_h2_stream_window = 64 * 1024;
+  uint32_t downstream_h2_conn_window = 256 * 1024;
+  uint32_t upstream_h2_stream_window = 512 * 1024;
+  uint32_t upstream_h2_conn_window = 1024 * 1024 * 1024; // Effectively unlimited
+
+  runTestWithBufferConfig(
+      {connection_watermark, downstream_h2_stream_window, downstream_h2_conn_window,
+       upstream_h2_stream_window, upstream_h2_conn_window},
+      num_requests, kNumBlocks, kBlockSize, num_requests * kNumBlocks * kBlockSize, kBlockSize);
+
+  EXPECT_LT(800, buffer_factory_->numBuffersCreated());
+  if (http2FlowControlV2()) {
+    EXPECT_LE(connection_watermark, buffer_factory_->maxBufferSize());
+    EXPECT_GE(2 * connection_watermark, buffer_factory_->maxBufferSize());
+    EXPECT_GT(2.0, std::get<0>(buffer_factory_->maxOverflowRatio()));
+  } else {
+    EXPECT_LE(50 * connection_watermark, buffer_factory_->maxBufferSize());
+    EXPECT_GE(70 * connection_watermark, buffer_factory_->maxBufferSize());
+    auto overflow_info = buffer_factory_->maxOverflowRatio();
+    EXPECT_LT(50.0, std::get<0>(overflow_info));
+    // Max overflow happens on the connection buffer.
+    EXPECT_EQ(connection_watermark, std::get<1>(overflow_info));
+  }
+  EXPECT_THAT(buffer_factory_->highWatermarkRange(), testing::Pair(32768, 512 * 1024));
+}
+
+// Verify that control frame protections take effect even if control frames end up queued internally
+// by nghttp2.
+TEST_P(Http2BufferWatermarksTest, PingFloodAfterHighWatermark) {
+  int num_requests = 1;
+  uint64_t kNumBlocks = 10;
+  uint64_t kBlockSize = 16 * 1024;
+  uint32_t connection_watermark = 32768;
+  uint32_t h2_stream_window = 64 * 1024;
+  uint32_t h2_conn_window = 1024 * 1024 * 1024; // Effectively unlimited
+
+  initializeWithBufferConfig(
+      {connection_watermark, h2_stream_window, h2_conn_window, h2_stream_window, h2_conn_window},
+      num_requests, true);
+
+  writev_matcher_->setWritevReturnsEgain();
+
+  sendRequests(num_requests, kNumBlocks, kBlockSize, connection_watermark + h2_stream_window);
+
+  floodServer(Http2Frame::makePingFrame(), "http2.outbound_control_flood",
+              ControlFrameFloodLimit + 1);
+}
+
+TEST_P(Http2BufferWatermarksTest, RstStreamWhileBlockedProxyingDataFrame) {
```

Comment:
I think the same tests for the case where upstream resets stream would be interesting and also stream timeout with pending DATA.

Reply:
Low level control of the upstream including manual parsing+serialization of H2 frames seems like a bit of a challenge.  We could consider enhancements to tests in a followup.

Reply:
We can probably make autonomous upstream reset streams on a specific DATA frame. Timeouts could be harder to test, maybe some special filter can trigger a timeout. Yes we can do it in a follow up.